test/test_ensemble.py

```diff
@@ -0,0 +1,94 @@
+from test.models.std_normal import StdNormal
+from bayes_kit.ensemble import AffineInvariantWalker
+import numpy as np
+import pytest as pt
+
+def run_sampling_test(sampler, model) -> None:
+    D = sampler._dim
+    K = sampler._num_walkers
+    M = 10000
+    draws = np.ndarray(shape=(M, K, D))
+    for m in range(M):
+        draws[m, 0:K, 0:D] = sampler.sample()
+    mean = np.mean(draws)
+    var = np.var(draws, ddof=1)
+    # sampler super inefficient with these settings, so need wide tolerance
+    # longer tests with M = 100_000 will converge much better but take several seconds
+    np.testing.assert_allclose(mean, model.posterior_mean(), atol=0.2)
+    np.testing.assert_allclose(var, model.posterior_variance(), atol=0.2)
+
+def test_aiw_exceptions() -> None:
+    model = StdNormal()
+    # illegal value: a
+    with pt.raises(ValueError):
+        sampler = AffineInvariantWalker(model, a = -1)
+    with pt.raises(ValueError):
+        sampler = AffineInvariantWalker(model, a = 0)
+    with pt.raises(ValueError):
+        sampler = AffineInvariantWalker(model, a = 1)
+
+    # illegal value: num_walkers
+    with pt.raises(ValueError):
+        sampler = AffineInvariantWalker(model, num_walkers = -1)
+    with pt.raises(ValueError):
+        sampler = AffineInvariantWalker(model, num_walkers = 0)
+    with pt.raises(ValueError):
+        sampler = AffineInvariantWalker(model, num_walkers = 1)
+
+    # illegal value: init
+    with pt.raises(ValueError):
+        sampler = AffineInvariantWalker(model, init = np.asarray([1.2, 2, 3]))
+
+    # illegal type: a
+    with pt.raises(TypeError):
+        sampler = AffineInvariantWalker(model, a = (1, 2, 3))
+        
+    # illegal type: seed
+    with pt.raises(TypeError):
+        sampler = AffineInvariantWalker(model, seed=1.234)
+
+    # illegal type: num_walkers
+    with pt.raises(TypeError):
+        sampler = AffineInvariantWalker(model, num_walkers = 2.39)
+
+    # illegal type: init
+    with pt.raises(TypeError):
+        sampler = AffineInvariantWalker(model, init = [1.2, 3.9])
+
+    # illegal type: model
+    model_dummy = "abc"
+    with pt.raises(AttributeError):
+        sampler = AffineInvariantWalker(model_dummy)
+
+        
+def test_aiw_std_normal() -> None:
```

Comment:
I'm not entirely sure what you're asking me to do here.  

> this can get a bit burdensome in test execution time

Our whole test suite, including this one, runs in 13s on my desktop.  Is this really a concern or can I just ignore this for now?


> it's not clear that the "sampler worked" base test is the right level of granularity
> If the run_sampling_test call passes with the sampler's default parameters, it isn't clear that rerunning it with different parameters can capture how those parameters have changed the underlying behavior. 

There's no good theory around this to test in the affine-invariant case.  The hyperparameter `a` is similar to the proposal scale in random-walk Metropolis.  There will be an optimal value.  In many cases we know what that is for Metropolis but there's not much theory around the affine-invariant ensemble methods.  Should we start by testing the behavior of differently scaled or differently shaped proposals in Metropolis?

If we're willing to open up the box and test internals, I could maybe rewrite parts of the code to expose as functions rather than just as code.  For example, we could test that the variance of proposals is higher with larger `a`.

> ssert on the dimensions of its _thetas member (along with maybe some tests for e.g. stretch_move specifically, to make sure that function is actually using all available walkers appropriately).

That won't test that they're being used appropriately.  Everything's stochastic all the way down in that updating a single walker involves selecting a random complementary walking, proposing, then doing an MH accept/reject step.  I'm not sure how to test things like that we have a uniform distribution over all complementary walkers.  Again, we'd have to decompose the sampler into finer grained pieces and test those.

> For another example, as written, we test that an initialization value can be passed successfully, but it's not clear whether that value was actually honored, e.g. by seeing whether choosing a good initial value results in faster convergence/tighter tolerance/etc.

Again, same problem for Metropolis and every other sampler that we use.  And again, we can open up the box and test that we really are setting the values appropriately on initialization.  We might want to test that everything breaks if each ensemble member is initialized at the same value.

Reply:
First off, this does not seem to me to be a blocker for the PR going through--as @roualdes says, it's definitely a good step forward!--and I don't intend to single out this PR for particular scrutiny, it just happened to be what made me realize that we need to revisit the topic of testing strategy.

Also, I wasn't clear enough in my earlier comment. I'm not requesting even-stricter tests; I just want to make sure that each test adds new information. If we have several tests that don't capture different specific behavior of the subject-under-test, one option is to write them to test more specific details; another is to say "These are testing the same thing, maybe we'd be fine cutting some of them to save runtime." Both are valid responses, as long as we're satisfied with the resulting coverage.

Here, I'm hearing you say that for the affine-invariant sampler, there's no good theory for how changing the hyperparameters should affect overall algorithmic performance. To me that sounds like it means that even if we rerun the convergence test, we don't have an outcome to assert on afterwards. So I'm thinking there's no need to spend time rerunning it--I'd be happy to cut entirely the tests against a specific initialization point, initializing with different random generators, etc.

(That said, *if* we do want to test those things, rerunning the full integration test for them will take longer & probably give less new information than testing something more specific. So for the initialization point, instead of repeating the test for convergence, I'd pick a start point out in the boonies with a model centered at 0, take one draw, and assert that the result is within $APPROPRIATE_VALUE of the initialization; for the random generators, pass in a mock generator and assert that that mock was used; etc.)

I do love the idea of testing for higher-order properties that any correct implementation ought to have--you mentioned ensuring correct distribution over walkers; hearing you mention it, that sounds like a great thing to check. For that specifically, we can mock `_stretch_move` with a function that captures the parameters it was called with, run 500 or so draws (which will be instantaneous, since we don't actually need to evaluate the densities), and then do some assertions on the distribution of parameters used to call it--I'm happy to help with setting this up, it'll be good for me too to figure out the mechanics. But again, I don't think that needs to be present for this PR to go through. (Similarly with checking for higher proposal variance with higher `a`--that sounds useful to add, and maybe something we can test either through a test against `_draw_z`, or with some clever mocking or a minor refactor, but I think that's more an "open an issue" thing than a blocker for the PR.)

> Our whole test suite, including this one, runs in 13s on my desktop. Is this really a concern or can I just ignore this for now?

Basically, it's a real concern that you can ignore for now :smile: I'm just looking at the possible matrix of parameters and thinking that even if each call to `run_sampling_tests` takes (say) half a second, the overall runtime could get noticeable. (And again, my objection to that is in the context where we don't have a good theoretical basis to test for new outcomes from the additional integration test calls--I'm perfectly sanguine about taking the extra time if it gives us key information, but it sounds like it might not here.)

> Should we start by testing the behavior of differently scaled or differently shaped proposals in Metropolis?

I think this gets into the tension between testing the algorithm and testing the implementation of the algorithm. For the purposes of testing the implementation, I don't think we need to go beyond testing that Metropolis calls what it's given; past that point, it seems like we'd be more testing the Metropolis algorithm itself rather than this implementation of it.

Reply:
I think this is the key issue that @jsoules raised:

> we'd be more testing the Metropolis algorithm itself rather than this implementation of it

I agree that we should test our implementations.  And that what we have now mainly tests the algorithm.  

We can test our implementation does the right thing with changes in a, namely that it gives you a broader range of proposals.  That's easy to check.  And that's why we were talking about mock objects on the way to lunch.  It does require refactoring the code so we can test the internal details of the algorithm.  Otherwise, all we have is something we can test end-to-end.  You can see that I just pushed the new tests for ESS and added the IAT functionality, where the IAT funcitonality is required for the ESS functionality.  We can do the same thing here.

Part of what I was doing with all those calls is making sure that we have code coverage.  The tests can be simplified.

Reply:
Sounds like we're on the same page!

Re: doing the right thing with changes in `a`, would you consider hitting `_draw_z` for N draws each with M different values of `a` (well, `_inv_sqrt_a` and `_sqrt_a`) to be sufficient? Or would you prefer to test only through "public" parts of the API?

My understanding was that the `z` interpolation factor was pretty specific to this sampler and so isn't likely to have outside consumers if promoted to a public API endpoint, so I'm comfortable leaving the code as currently structured and just testing the private function if it avoids some of the refactoring. Also happy to go along with a bigger restructure if there's a practical or policy reason to do so--just looking for an appropriately minimized amount of effort here.

(To be clear, I'm imagining a test that in pseudocode says basically:
```python
def test_variance_of_interpolation_factor_increases_as_a_increases():
    def test_an_a(a: int) -> Sample[]:
        sampler = # instantiate an AffineInvariantWalker with the given a
        return [sampler._draw_z() for _ in range(100)]
    as = [1, 2, 5, 10, 100]
    zs = [test_an_a(a) for a in as]
    # Do something to fix dimensions, then
    assert var(z[0]) < var(z[1]) < var(z[2]) < var(z[3]) < var(z[4])  
```
Naturally we could get a lot more specific about the statistical properties we'd expect from `_draw_z()` and the tolerance to which we'd want those to be realized, this is just what I had in mind right now.)